lib/api/units/validateBody.js

```diff
@@ -0,0 +1,233 @@
+      }`,
+      severity: 'error'
+    };
+
+    return [error, null];
+  }
+}
+
+/**
+ * Returns a body validator class based on the given
+ * real and expected body media types.
+ * @param {MediaType} realType
+ * @param {MediaType} expectedType
+ * @returns {Validator}
+ */
+function getBodyValidator(realType, expectedType) {
+  const both = (predicate) => (real, expected) => {
+    return [real, expected].every(predicate);
+  };
+
+  const validators = [
+    [TextDiff, both(isPlainText)],
+    // List JsonSchema first, because weak predicate of JsonExample
+    // would resolve on "application/schema+json" media type too.
+    [
+      JsonSchema,
+      (real, expected) => {
+        return (
+          isJson(real) &&
+          expected.type === 'application' &&
+          expected.subtype === 'schema' &&
+          expected.suffix === 'json'
+        );
+      }
+    ],
+    [JsonExample, both(isJsonWeak)]
+  ];
+
+  const validator = validators.find(([_name, predicate]) => {
+    return predicate(realType, expectedType);
+  });
+
+  if (!validator) {
+    const error = {
+      message: `Can't validate real media type '${mediaTyper.format(
+        realType
+      )}' against expected media type '${mediaTyper.format(expectedType)}'.`,
+      severity: 'error'
+    };
+    return [error, null];
+  }
+
+  return [null, validator[0]];
+}
+
+/**
+ * Validates given bodies of transaction elements.
+ * @param {Object} real
+ * @param {Object} expected
+ */
+function validateBody(real, expected) {
+  const results = [];
+  const bodyType = typeof real.body;
+
+  if (bodyType !== 'string') {
+    throw new Error(`Expected HTTP body to be a String, but got: ${bodyType}`);
+  }
+
+  const [realTypeError, realType] = getBodyType(real.body, real.headers);
+  const [expectedTypeError, expectedType] = expected.bodySchema
+    ? getBodySchemaType(expected.bodySchema)
+    : getBodyType(expected.body, expected.headers);
+
+  if (realTypeError) {
+    results.push(realTypeError);
+  }
+
+  if (expectedTypeError) {
+    results.push(expectedTypeError);
+  }
+
+  const hasErrors = results.some((result) =>
+    ['error'].includes(result.severity)
+  );
+
+  // Skipping body validation in case errors during
+  // real/expected body type definition.
+  const [validatorError, ValidatorClass] = hasErrors
+    ? [null, null]
+    : getBodyValidator(realType, expectedType);
+
+  if (validatorError) {
+    results.push(validatorError);
+  }
+
+  const usesJsonSchema = ValidatorClass && ValidatorClass.name === 'JsonSchema';
+  const validator =
+    ValidatorClass &&
+    new ValidatorClass(
+      real.body,
+      usesJsonSchema ? expected.bodySchema : expected.body
+    );
+  const rawData = validator && validator.validate();
+  const validatorResults = validator ? validator.evaluateOutputToResults() : [];
+  results.push(...validatorResults);
+
+  return {
+    validator: ValidatorClass && ValidatorClass.name,
+    realType: mediaTyper.format(realType),
```

Comment:
Vice versa: they are kept as parsed to prevent additional parsing, and formatted into a string in the validation result to retain the same output format.

Reply:
Yes, I'm sorry, I didn't formulate the comment well. I did see the flow, but wasn't sure why it is needed, i.e. why it would need to be parsed twice. Now I noticed the type is needed as a parameter for the `getBodyValidator(realType)` where it would need to be parsed twice and suddenly it all clicked 👍